docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,309 @@
+# Automatic reachable services
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6551
+
+## Context and Problem Statement
+
+With large enough Mesh (more than 1000 services) processing the whole cluster state for each proxy is not possible due to performance reasons. 
+The complexity of this computation is O(n^2) (n is the number of services), it also affects Envoy configs size (reaching multiple MB).
+This is why a feature called [reachable services](https://github.com/kumahq/kuma-website/blob/045ff9868acdb4b322759dde3f8053f41948100f/app/_src/production/upgrades-tuning/fine-tuning.md#L8-L7) was introduced.
+Reachable services is a step in the right direction but defining the whole service graph manually by the service owners or the platform team is a big ask (it can be another reason to resist adopting the mesh).
+
+How can we make Kuma easy to use and not have people needing to do systematic performance tuning?
+
+## Glossary
+
+- MTP - [MeshTrafficPermission](https://kuma.io/docs/2.4.x/policies/meshtrafficpermission/#meshtrafficpermission-(beta))
+- ODCDS - [On-demand CDS Updates](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/on_demand_updates_filter#config-http-filters-on-demand)
+
+## Decision Drivers 
+
+* feasibility of implementation for 2.5 release
+* minimize the manual labour needed
+* smooth migration to Mesh for large systems without downtime or service communication issues
+
+## Considered Options
+
+* Using MeshTrafficPermission to generate reverse graph that would feed into reachable services
+* Dynamically load clusters using [ODCDS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/on_demand/v3/on_demand.proto)
+* Using MeshHTTPRoute / MeshTCPRoute to populate the clusters
+
+## Decision Outcome
+
+Chosen option: "Using MeshTrafficPermission to generate reverse graph that would feed into reachable services"
+
+### Positive Consequences
+
+* not invasive, composable, plug and play, can be removed if proven not efficient enough
```

Comment:
I have a question related to the ExternalServices and their interaction with the 'reachableServices' configuration. Currently, it appears that specifying ExternalServices in the 'reachableServices' doesn't have an effect. However, I'm curious if doing so could offer advantages in terms of security, configuration management, or performance optimization. Also, if we would like to do it how would that interact with autorechable?

Reply:
Ok, maybe I haven't noticed this. When reachableServices don't specify ExternalService's name it isn't available. If that works it's fine to me. I've checked the code and thought that it might not affect them

Reply:
Right now auto-reachable-services are combined with rechableservices, just like mentioned in - https://github.com/kumahq/kuma/pull/8084#discussion_r1371523549 - but @jakubdyszkiewicz wanted to change that.

Reply:
Also, does it relate to your change: https://github.com/kumahq/kuma/issues/6589

Reply:
I think it does, if you don't define traffic permission for ES it won't get the cluster. @jakubdyszkiewicz do we want to treat ES differently?

Reply:
No, we treat it as any other service

Reply:
Resolving conversation since Lukasz approved this PR (I guess this answer was enough). Please reopen otherwise.